File: ts-consign-service/app/security/jwt.py
from __future__ import annotations

import base64
import hashlib
import hmac
import json
import time
from dataclasses import dataclass


@dataclass(frozen=True)
class Authentication:
    present: bool
    roles: frozenset[str]


def parse_optional_bearer(authorization: str | None, now: float | None = None) -> Authentication:
    if not authorization or not authorization.startswith("Bearer "):
        return Authentication(False, frozenset())
    token = authorization[7:]
    try:
        header_part, payload_part, signature_part = token.split(".")
        header = json.loads(_decode(header_part))
        payload = json.loads(_decode(payload_part))
        if header.get("alg") != "HS256":
            raise ValueError("unsupported JWT")
        expected = hmac.new(b"secret", f"{header_part}.{payload_part}".encode(), hashlib.sha256).digest()
        signature = base64.urlsafe_b64decode(_pad(signature_part))
        if not hmac.compare_digest(signature, expected):
            raise ValueError("invalid signature")
        if float(payload["exp"]) * 1000 < (now or time.time()) * 1000:
            raise ValueError("expired JWT")
        roles = payload.get("roles")
        if not isinstance(roles, list):
            raise ValueError("invalid roles")
        return Authentication(True, frozenset(str(role) for role in roles))
    except (KeyError, TypeError, ValueError, json.JSONDecodeError) as exc:
        raise ValueError("invalid JWT") from exc


def _decode(value: str) -> str:
    return base64.urlsafe_b64decode(_pad(value)).decode()


def _pad(value: str) -> str:
    return value + "=" * (-len(value) % 4)

File: ts-station-service/internal/security/jwt.go
package security

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"strings"
	"time"
)

type Authentication struct {
	Present bool
	Roles   map[string]bool
}

func ParseOptionalBearer(authorization string, now time.Time) (Authentication, error) {
	if !strings.HasPrefix(authorization, "Bearer ") {
		return Authentication{}, nil
	}
	parts := strings.Split(strings.TrimPrefix(authorization, "Bearer "), ".")
	if len(parts) != 3 {
		return Authentication{}, errors.New("malformed JWT")
	}
	headerBytes, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil {
		return Authentication{}, err
	}
	var header struct {
		Algorithm string `json:"alg"`
	}
	if err := json.Unmarshal(headerBytes, &header); err != nil || header.Algorithm != "HS256" {
		return Authentication{}, errors.New("unsupported JWT")
	}
	signature, err := base64.RawURLEncoding.DecodeString(parts[2])
	if err != nil {
		return Authentication{}, err
	}
	mac := hmac.New(sha256.New, []byte("secret"))
	mac.Write([]byte(parts[0] + "." + parts[1]))
	if !hmac.Equal(signature, mac.Sum(nil)) {
		return Authentication{}, errors.New("invalid JWT signature")
	}
	payloadBytes, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return Authentication{}, err
	}
	var payload struct {
		Exp   json.Number `json:"exp"`
		Roles []string    `json:"roles"`
	}
	decoder := json.NewDecoder(strings.NewReader(string(payloadBytes)))
	decoder.UseNumber()
	if err := decoder.Decode(&payload); err != nil {
		return Authentication{}, err
	}
	if payload.Exp == "" {
		return Authentication{}, errors.New("JWT has no expiration")
	}
	expiration, err := payload.Exp.Int64()
	if err != nil {
		return Authentication{}, err
	}
	if time.Unix(expiration, 0).Before(now) {
		return Authentication{}, errors.New("JWT expired")
	}
	roles := make(map[string]bool, len(payload.Roles))
	for _, role := range payload.Roles {
		roles[role] = true
	}
	return Authentication{Present: true, Roles: roles}, nil
}
